FOCUS #181: Usage Unit Dimension (#184)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/dimensions/usageunit.md

# Usage Unit

Usage Unit is the unit of measure for the usage or consumption of a resource or service. Usage Unit is used in conjunction with Quantity In Usage Unit to gain understanding of usage or consumption of a given product or service in an environment.

The UsageUnit column MUST be present in the billing data. This column MUST be of type String and MUST NOT contain null values when the ChargeType is 'Usage'. UsageUnit should be expressed as a single unit of measure.  Units of measure used in UsageUnit SHOULD adhere to the values and format requirements specified in the Recommended Values section below.  The UsageUnit column MUST not be used as the basis for determining values related to any pricing or cost metric.

## Column ID

UsageUnit

## Display name

Usage Unit

## Description

The unit of measure for the usage or consumption of a resource or service

> The Usage Unit for a particular line item may differ from PricingUnit when providers use different units, unit increments, or columns to determine cost. The values in UsageUnit and QuantityInUsageUnit are often listed at a finer granularity or over a different time interval than the PricingUnit and QuantityInPricingUnit. While UsageUnit may appear to relate to pricing and cost, UsageUnit is focused on resource or service consumption.  It is essential not to confuse UsageUnit with PricingUnit which is the basis for determining cost.

Reviewer: _The Usage Unit for a particular line item may differ from PricingUnit when providers use different** units, unit increments, or **columns to determine cost._

Could you consider rephrasing this? Referring to this part in particular: _...**when providers use different ... columns to determine cost**..._

> The Usage Unit for a particular line item may differ from PricingUnit when providers use different units, unit increments, or columns to determine cost. The values in UsageUnit and QuantityInUsageUnit are often listed at a finer granularity or over a different time interval than the PricingUnit and QuantityInPricingUnit. While UsageUnit may appear to relate to pricing and cost, UsageUnit is focused on resource or service consumption.  It is essential not to confuse UsageUnit with PricingUnit which is the basis for determining cost.

Reviewer 2: Style wise, we should stick to calling it `Usage Unit` or `UsageUnit` consistently . It makes sense to refer to the human readable name in the intro, but going around the horn, it feels like we refer to the column name after that.

```suggestion
The UsageUnit for a particular line item may differ from PricingUnit when providers use different units, unit increments, or columns to determine cost. The values in UsageUnit and QuantityInUsageUnit are often listed at a finer granularity or over a different time interval than the PricingUnit and QuantityInPricingUnit. While UsageUnit may appear to relate to pricing and cost, UsageUnit is focused on resource or service consumption.  It is essential not to confuse UsageUnit with PricingUnit, which is the basis for determining cost.
```

Reviewer 3: Please use **PricingMeasure** (non-normalized) instead of PricingUnit (semi-normalized)

### Recommended Values

> Usage units SHOULD be expressed as a single unit of measure adhering to one of the following three formats.

Reviewer 2: ```suggestion
UsageUnit SHOULD be expressed as a single unit of measure adhering to one of the following three formats.
```

* `<plural-units>` - "GB", "Seconds"
* `<singular-unit> <plural-time-units>` - "GB Hours", "MB Days"
* `<plural-units>/<singular-time-unit>` - "GB/Hour", "PB/Day"

> Usage units MAY be expressed with a unit quantity or time interval.  If a unit quantity or time interval is used, the unit quantity or time interval MUST be expressed as a whole number.  The following formats are valid.

Reviewer 2: ```suggestion
UsageUnit MAY be expressed with a unit quantity or time interval.  If a unit quantity or time interval is used, the unit quantity or time interval MUST be expressed as a whole number.  The following formats are valid:
```

* `<quantity> <plural-units>` - "1000 Tokens", "1000 Characters"
* `<plural-units>/<interval> <plural-time-units>` - "Units/3 Months"

### Recommended Usage Unit Values

> Unit values and components of the UnitName column MUST use a capitalization scheme that is consistent with the capitalization scheme used in this section if that term is listed in this section. For example, a value of "gigabyte seconds" would not be compliant with this specification as the terms "gigabyte" and "second" are listed in this section with the appropriate capitalization.  If the unit is not listed in the table, it is to be used over a functional equivalent with similar meaning.

Reviewer 2: ```suggestion
Unit values and components of the UsageUnit column MUST use a capitalization scheme that is consistent with the capitalization scheme used in this section, if that term is listed in this section. For example, a value of "gigabyte-seconds" would not be compliant with this specification as the terms "gigabyte" and "second" are listed in this section with the appropriate capitalization.  If the unit is not listed in the table, it is to be used over a functional equivalent with similar meaning.
```

> UsageUnit SHOULD be composed of the list of recommended units listed in this section unless the UsageUnit value covers a dimension not listed in the recommended unit set or if the unit covers a count-based unit distinct from recommended values in the count dimension listed in this section.  

Reviewer 2: ```suggestion
UsageUnit SHOULD be composed of the list of recommended units listed in this section, unless the UsageUnit value covers a dimension not listed in the recommended unit set, or if the unit covers a count-based unit distinct from recommended values in the count dimension listed in this section.  
```

| Data size in bits    | Data size in bytes    |
| -------------------- | --------------------- |
| Kib (kibibit = 2^10) | KiB (kibibyte = 2^10) |
| Mib (mebibit = 2^20) | MiB (mebibyte = 2^20) |
| Gib (gibibit = 2^30) | GiB (gibibyte = 2^30) |
| Tib (tebibit = 2^40) | TiB (tebibyte = 2^40) |
| Pib (pebibit = 2^50) | PiB (pebibyte = 2^50) |
| Eib (exbibit = 2^60) | EiB (exbibyte = 2^60) |
| Kb (kilobit = 10^3)  | KB (kilobyte = 10^3)  |
| Mb (megabit = 10^6)  | MB (megabyte = 10^6)  |
| Gb (gigabit = 10^9)  | GB (gigabyte = 10^9)  |
| Tb (terabit = 10^12) | TB (terabyte = 10^12) |
| Pb (petabit = 10^15) | PB (petabyte = 10^15) |
| Eb (exabit = 10^15)  | EB (exabyte = 10^15)  |

#### Count-based Unit Names

A count-based unit is a noun that represents a discrete number of items, events, or services.  For example, a count-based unit can be used to represent the number of requests, instances, tokens, or connections.  

> If the list of recommended values does not cover a count-based unit, a provider may introduce a new noun representing a count-based unit.  All nouns appearing in UsageUnit that are not listed in the recommended values table will be considered count-based units.  A new count-based unit value MUST be capitalized.

Reviewer 2: ```suggestion
If the list of recommended values does not cover a count-based unit, a provider MAY introduce a new noun representing a count-based unit.  All nouns appearing in UsageUnit that are not listed in the recommended values table will be considered count-based units.  A new count-based unit value MUST be capitalized.
```

| Count        |
|--------------|
| Count        |
| Unit         |
| Request      |
| Token        |
| Connection   |
| Certificate  |
| Domain       |
| Core         |

#### Time-based Unit Names

A time-based unit is a noun that represents a time interval.  Time-based units can be used to measure consumption over a time interval or in combination with another unit to capture a rate of consumption.

| Time         |
|--------------|
| Year         |
| Month        |
| Day          |
| Hour         |
| Minute       |
| Second       |

### Composite Units

If the UsageUnit value is a composite value made from combinations of one or more units, each component MUST also align with the set of recommended values.

> Instead of "per" or "-" to denote a Composite Unit, slash ("/") and space(" ") must be used as a common convention.  Count-based units like requests, instances, and tokens SHOULD be expressed using a value listed in the count dimension.  For example, if a usage unit is measured as a rate of requests or instances over a period of time, the unit should be listed as "count/day" to signify the number of requests per day.

Reviewer 2: ```suggestion
Instead of "per" or "-" to denote a Composite Unit, slash ("/") and space(" ") MUST be used as a common convention.  Count-based units like requests, instances, and tokens SHOULD be expressed using a value listed in the count dimension.  For example, if a usage unit is measured as a rate of requests or instances over a period of time, the unit SHOULD be listed as "count/day" to signify the number of requests per day.
```

# Usage Unit

Usage Unit represents the units of a given resource and/or service used or purchased in combination with [Usage Quantity](#usagequantity). Usage Unit is often listed at a finer granularity or over a different time interval when compared to the [Pricing Unit](#pricingunit) (complementary to [Pricing Quantity](#pricingquantity)), and focuses on resource and service consumption, not pricing and cost.

> The UsageUnit column MUST be present in the billing data. This column MUST be of type String and MUST NOT contain null values when the ChargeType is 'Usage'. Units of measure used in UsageUnit SHOULD adhere to the values and format requirements specified in the [Unit Format](#unitformat) attribute. The UsageUnit column MUST NOT be used as the basis for determining values related to any pricing or cost metric.

Reviewer: ```suggestion
The UsageUnit column MUST be present in the billing data. This column MUST be of type String and MUST NOT contain null values when the ChargeType is 'Usage'. Units of measure used in UsageUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute. The UsageUnit column MUST NOT be used as the basis for determining values related to any pricing or cost metric.
```

## Column ID

UsageUnit

## Display name

Usage Unit

## Description

Units of a given resource and/or service used or purchased in combination with [Usage Quantity](#usagequantity).

## Content constraints

|    Constraint   |      Value      |
|:----------------|:----------------|
| Column required | True            |
| Data type       | String          |
| Allows nulls    | True            |

> | Value format    | Unit Format     |

Reviewer: ```suggestion
| Value format    | Unit Format recommended |
```

Reviewer: This suggestion is based on the normative paragraph where we stated that it `SHOULD adhere to...`
If you choose a different term, please keep in mind that we need to address this in the PricingUnit spec as well.

# Usage Unit

Usage Unit refers to a unit of measurement for the consumption or usage of resources or services. The Usage Unit for a particular line item may differ from the published Pricing Unit when providers use different units, unit increments, or dimensions for calculating cost.  This unit is commonly used when auditing or reconciling consumption or billing data.

> The UsageUnit column MUST be present and MUST NOT be null or empty. UsageUnit is of type String, and UsageUnit MUST be composed of the list of recommended units or unit abbreviations if it is possible to be expressed with the units listed in "Recommended Unit Set" below. If the UsageUnit value is a composite value made from combinations of one or more units, each component MUST also align with the set of recommended units if possible.  For example, a composite unit of "GB/Month" is allowed.

Reviewer 4: If we are saying it MUST if possible - wouldn't it be more concise to say SHOULD?

Reviewer 5: Agreed, I tried to preserve the language of the original, but I agree that is an odd contradiction, you "MUST" do this if it's possible.  I think we need to change it into something more along the lines of:

Reviewer 5: @MichaelArkoosh-vega - see the PR I just opened against this branch.

# Usage Unit

> Usage Unit refers to a unit of measurement for the purchase or consumption of resources or services. The Usage Unit for a particular line item may differ from the published Pricing Unit when providers use different units, unit increments, or dimensions for calculating cost.  This unit is commonly used when auditing or reconciling consumption or billing data.

Reviewer 5: Purchase was in there at the recommendation of @flanakin - Michael?

The UsageUnit column MUST be present and MUST NOT be null or empty. UsageUnit is of type String. UsageUnit should be expressed as a single unit of measure adhering to the following format options:

* &lt;plural-units&gt; - "GB", "Seconds"
* &lt;singular-unit&gt; &lt;plural-time-units&gt; - "GB Hours", "MB Days"
* &lt;plural-units&gt;/&lt;singular-time-unit&gt; - "GB/Hour", "PB/Day"

> 

Reviewer 5: I moved it down to Allowed Values.  The section now leads with this, but it doesn't add BCF-14 language to the requirement.

> If the UsageUnit value is a composite value made from combinations of one or more units, each component MUST also align with the set of allowed values.

Reviewer 6: What's the plan when the unit of measurement is not time, data or count? E.g. what if it's a distance for a particular type of service or something different? How do we ensure we're not saying it MUST be one of these 3 we've specified? If we want the spec to apply to cloud, saas and other things, it seems we can only recommend?

Reviewer 5: See the latest update - I modified the working to be a bit more explicit on dimensions.

> UsageUnit MUST be composed of the list of recommended units listed in "Allowed Values" below unless the UsageUnit value covers a dimension not listed in the recommended unit set or if the unit covers a count-based unit distinct from allowed values in the count dimension listed in "Allowed Values."  

Reviewer 5: Done

| Data size in bits    | Data size in bytes    |
| -------------------- | --------------------- |
| Kib (kibibit = 2^10) | KiB (kibibyte = 2^10) |
| Mib (mebibit = 2^20) | MiB (mebibyte = 2^20) |
| Gib (gibibit = 2^30) | GiB (gibibyte = 2^30) |
| Tib (tebibit = 2^40) | TiB (tebibyte = 2^40) |
| Pib (pebibit = 2^50) | PiB (pebibyte = 2^50) |
| Eib (exbibit = 2^60) | EiB (exbibyte = 2^60) |
| Kb (kilobit = 10^3)  | KB (kilobyte = 10^3)  |
| Mb (megabit = 10^6)  | MB (megabyte = 10^6)  |
| Gb (gigabit = 10^9)  | GB (gigabyte = 10^9)  |
| Tb (terabit = 10^12) | TB (terabyte = 10^12) |
| Pb (petabit = 10^15) | PB (petabyte = 10^15) |
| Eb (exabit = 10^15)  | EB (exabyte = 10^15)  |

### Count-based Units

A count-based unit is a noun that represents a discrete number of items, events, or services.  For example, a count-based unit can be used to represent the number of requests, instances, tokens, or connections.  If the list of allow values does not cover a count-based unit, a provider may introduce a new noun representing a count-based unit.  All nouns appearing in UsageUnit that are not listed in the allowed values table will be considered count-based units.  

> Any new count-based units introduced MUST use a capitalization scheme that is consistent with the capitalization scheme used in the allowed values table.  For example, if a provider introduces a new count-based unit "Thing", the capitalization scheme MUST be "Thing" and not "thing" or "Thing".

Reviewer 5: I blame autocorrect, it was supposed to be "THING" - resolved in next push

# Usage Unit

> Usage Unit refers to a unit of measurement for the consumption of resources or services. The Usage Unit for a particular line item may differ from the published Pricing Unit when providers use different units, unit increments, or dimensions for calculating cost.  This unit is commonly used when auditing or reconciling consumption data.

Reviewer: At this point I don't have a suggestion for improvement :(, but this reference/comparison to Pricing Unit seems a bit unclear. It implies that there are UCs in which providers use different columns for cost calculation? The prescribed calculation is QuantityInPricingUnit multiplied by the corresponding unit price.

Reviewer 5: I know, right?  I revised it form a previous version.  I can see the reasoning - I know of one provider that uses different "usage" units than "pricing" units.   Let's seek some input form @arrarama and @flanakin here

Reviewer 5: Discussed in the VS3 meeting with a number of folks - I'm going to add language to make it crystal clear that this shouldn't be used for pricing calc

Reviewer 5: I'll open up another discussion against the commits I just made to add in one clarifying sentence.

The UsageUnit column MUST be present in the billing data. This column MUST be of type String and MUST NOT contain null values when the ChargeType is 'Usage'. UsageUnit should be expressed as a single unit of measure adhering to the values and format requirements specified in the Allowed Values section below.

## Column ID

UsageUnit

## Display name

Usage Unit

## Description

A unit of measurement for the consumption of resources or services

## Content constraints

|    Constraint   |      Value      |
|:----------------|:----------------|
| Column required | True            |
| Data type       | String          |
| Allows nulls    | True            |
| Value format    | list-of-values |

## Allowed Values

Usage units MUST be expressed as a single unit of measure adhering one of the following three formats.

* &lt;plural-units&gt; - "GB", "Seconds"
* &lt;singular-unit&gt; &lt;plural-time-units&gt; - "GB Hours", "MB Days"

> * &lt;plural-units&gt;/&lt;singular-time-unit&gt; - "GB/Hour", "PB/Day"

Reviewer 5: Awesome, thanks - didn't even know that was possible.  Thanks for the typo catch above as well

> 

Reviewer 5: I"m OK with this, maybe let's ask in this week's call.

# Usage Unit

> Usage Unit is a measure of the amount of a resource or service was used or purchased. The Usage Unit for a particular line item may differ from the published Pricing Unit when providers use different units, unit increments, or columns for calculating cost. The UsageUnit column should not be used to calculate metrics related to any cost metric. This unit is commonly used when auditing or reconciling usage data.

Reviewer 5: @ijurica - I added the sentence "The UsageUnit column should not be used to calculate metrics related to any cost metric."

Reviewer: Similar cross-references will be included in PricingUnit and QuantityInPricingUnit specs. 

Couple of comments/suggestions available in one of the conversations in [PricingUnit PR](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/179) and VS1 team scheduled a 30 min discussion on this topic for Friday, at14:30 - 15:00 UTC. I'm well aware that this is a late notice. It would be great if you could join us.

# Usage Unit

Usage Unit is the unit of measure for the usage or consumption of a resource or service. Usage Unit is used in conjunction with Quantity In Usage Unit to gain understanding of usage or consumption of a given product or service in an environment.

> The UsageUnit column MUST be present in the billing data. This column MUST be of type String and MUST NOT contain null values when the ChargeType is 'Usage'. UsageUnit should be expressed as a single unit of measure.  Units of measure used in UsageUnit SHOULD adhere to the values and format requirements specified in the Recommended Values section below.

Reviewer 6: I think here we have to go MUST and then in the specific requirements, we have some SHOULDs, some MUSTs and some MAYs. If we go SHOULD here, a MUST in the below section becomes optional? This is the first time we're running into this so check my logic and accept only if it makes sense.

```suggestion
The UsageUnit column MUST be present in the billing data. This column MUST be of type String and MUST NOT contain null values when the ChargeType is 'Usage'. UsageUnit should be expressed as a single unit of measure.  Units of measure used in UsageUnit MUST adhere to the values and format requirements specified in the Recommended Values section below.
```

> The Usage Unit for a particular line item may differ from PricingUnit when providers use different units, unit increments, or columns to determine cost. The values in UsageUnit and QuantityInUsageUnit are often listed at a finer granularity or over a different time interval than the PricingUnit and QuantityInPricingUnit. While UsageUnit may appear to relate to pricing and cost, UsageUnit is focused on resource or service consumption.  It is essential not to confuse UsageUnit with PricingUnit which is the basis for determining cost. The UsageUnit column MUST not be used as the basis for determining values related to any pricing or cost metric.

Reviewer 6: nit: In the past, we've tried to keep all the normative content in a single paragraph (or bulleted list) usually towards the end. This has normative content in the middle and also the last sentence of the final paragraph. Just calling out in case we think its worth making consistent. I don't feel strongly it MUST change but if we can change and keep the same messaging, we should.

Reviewer 7: Agree with Udam here.  I would actually translate this into an example to show how they are different rather than declaring such ... ~

"""
Typically, PricingUnit and UsageUnit may appear synonymous but not always.  Below are a few examples showing ... 
"""

## Column ID

UsageUnit

## Display name

Usage Unit

## Description

The unit of measure for the usage or consumption of a resource or service

## Content constraints

|    Constraint   |      Value      |
|:----------------|:----------------|
| Column required | True            |
| Data type       | String          |
| Allows nulls    | True            |
| Value format    | \<not specified> |

## Recommended Values

Usage units SHOULD be expressed as a single unit of measure adhering to one of the following three formats.

* `<plural-units>` - "GB", "Seconds"

> * `<singular-unit> <plural-time-units>` - "GB Hours", "MB Days"

Reviewer 7: AWS and GCP use `<singular-unit>-<plural-time-units>` (with a dash separator).  Are we deviating for a reason?

AWS: https://repost.aws/knowledge-center/ebs-volume-charges
GCP: https://cloud.google.com/storage/pricing#price-tables

Reviewer 8: ```suggestion
* `<singular-unit>-<plural-time-units>` - "GB Hours", "MB Days"
```

Discussed on VS3 call and aligned on use of dash.

* `<plural-units>/<singular-time-unit>` - "GB/Hour", "PB/Day"

Usage units MAY be expressed with a unit quantity or time interval.  If a unit quantity or time interval is used, the unit quantity or time interval MUST be expressed as a whole number.  The following formats are valid.

* `<quantity> <plural-units>` - "1000 Tokens", "1000 Characters"
* `<plural-units>/<interval> <plural-time-units>` - "Units/3 Months"

UsageUnit SHOULD be composed of the list of recommended units listed in this section unless the UsageUnit value covers a dimension not listed in the recommended unit set or if the unit covers a count-based unit distinct from recommended values in the count dimension listed in this section.  

> Unit names are listed in this section with the appropriate capitalization.  If the unit is not listed in the table, it is to be used over a functional equivalent with similar meaning or incompatible capitalization.  

Reviewer 6: I think the sentences here need some BCP-14 terms (SHOULD?) right?

I'm also not sure the last sentence clearly states the requirement ("it is to be used over a functional equivalent"). Could be my bad english :)

Reviewer 8: Discussed on VS3 call.  @tobrien, we should discuss ASAP (next week's call at latest) to resolve.

> The following table and next section list the valid singular units for 3 dimensions of measurement: time, data, and count.  

Reviewer 6: Should we just split the table and have 3 sub headings here (you already have two, just make a Time Units and put the values in the respective sections. Seems clearer.

Reviewer 8: @tobrien Same comment as above; discuss soon.

| Hour         | Token        |
| Minute       | Connection   |
| Second       | Certificate  |
|              | Domain       |
|              | Core         |

### Data Size Units

Data size MUST be abbreviated using the following standard abbreviations.  Each data size abbreviation can be considered both the singular and plural form of the unit.  For example, "GB" is both the singular and plural form of the unit "GB".  The following table lists the valid data size units.

| Data size in bits    | Data size in bytes    |
| -------------------- | --------------------- |
| Kib (kibibit = 2^10) | KiB (kibibyte = 2^10) |
| Mib (mebibit = 2^20) | MiB (mebibyte = 2^20) |
| Gib (gibibit = 2^30) | GiB (gibibyte = 2^30) |
| Tib (tebibit = 2^40) | TiB (tebibyte = 2^40) |
| Pib (pebibit = 2^50) | PiB (pebibyte = 2^50) |
| Eib (exbibit = 2^60) | EiB (exbibyte = 2^60) |
| Kb (kilobit = 10^3)  | KB (kilobyte = 10^3)  |
| Mb (megabit = 10^6)  | MB (megabyte = 10^6)  |
| Gb (gigabit = 10^9)  | GB (gigabyte = 10^9)  |
| Tb (terabit = 10^12) | TB (terabyte = 10^12) |
| Pb (petabit = 10^15) | PB (petabyte = 10^15) |
| Eb (exabit = 10^15)  | EB (exabyte = 10^15)  |

### Count-based Units

> A count-based unit is a noun that represents a discrete number of items, events, or services.  For example, a count-based unit can be used to represent the number of requests, instances, tokens, or connections.  If the list of recommended values does not cover a count-based unit, a provider may introduce a new noun representing a count-based unit.  All nouns appearing in UsageUnit that are not listed in the recommended values table will be considered count-based units.  

Reviewer 6: ```suggestion
A count-based unit is a noun that represents a discrete number of items, events, or services.  For example, a count-based unit can be used to represent the number of requests, instances, tokens, or connections.  If the list of recommended values does not cover a count-based unit, a provider MAY introduce a new noun representing a count-based unit.  All nouns appearing in UsageUnit that are not listed in the recommended values table will be considered count-based units.  
```
